### ISSUE TITLE
Long press doesn't open context menu in various parts of Zotero on Windows with a touchscreen.

### ISSUE THREAD

**Issue by mrtcode** [zotero/zotero#4094]
https://forums.zotero.org/discussion/108057/zotero-7-beta-touch

**Comment by abaevbog** [comment]
I can reproduce this. There is a square appearing in the long press area and then - nothing.

If I do a long press elsewhere (e.g. file explorer), the context menu appears after the long press is released and the square goes away. I will try to take advantage of having a (albeit barely working) touchscreen windows to get some info on this.

**Comment by abaevbog** [comment]
> I would guess this comes down to places where we're handling contextmenu ([1](https://github.com/zotero/zotero/blob/89320a7ee4d30b0ae35763e742d31e54cb6c615f/chrome/content/zotero/elements/tagsBox.js#L65)) vs click/mousedown ([2](https://github.com/zotero/zotero/blob/2f4a232c41aa3955bba093828b71b9aadc44a21d/chrome/content/zotero/components/virtualized-table.jsx#L725)), but I don't have a touchscreen Windows machine to test with.

It looks like that is exactly the reason why. Long press only dispatches `contextmenu` event without the `mousedown` event as with right-click. So in cases where `contextmenu` events are not handled (like the virtualized table nodes), long press does nothing.

One way to address this would be to just add a dedicated `contextmenu` listener and move the popup opening logic there from the `mousedown` handlers. Is there a reason for why some context menus have to be handled in `mousedown` listeners?

Alternatively, we can leave `mousedown` handlers as is and just call them when `contextmenu` event is picked up. For example, adding this to [_renderItem](https://github.com/zotero/zotero/blob/85f58cc97d2b73af7777c2d6e632dc7b2631531c/chrome/content/zotero/components/virtualized-table.jsx#L1146-L1152) makes the context menu appear on the long press as well.

```
node.addEventListener('contextmenu', (e) => this._handleMouseDown({ button: 2, screenX: e.screenX, screenY: e.screenY }, index), { passive: true });
```

**Comment by abaevbog** [comment]
> I'm not personally a fan of creating fake events when we could just extract/move the logic instead.

Fully agree - I only brought up the alternative in case if there are instances where, maybe, `contextmenu` event wasn't firing properly, so handling in `mousedown` was necessary? Or other reasons like this that I may not be fully aware of. Because if there are no such reasons, I feel like extracting the popup-opening logic into a separate listener shouldn't be that bad...


> What if we modify _handleMouseDown in the future and add a dependency on some property that we aren't setting in that fake event object?

So something along the lines of `if (e.button == 2 || e.detail.longpressContextMenu)` [here](https://github.com/zotero/zotero/blob/89320a7ee4d30b0ae35763e742d31e54cb6c615f/chrome/content/zotero/components/virtualized-table.jsx#L721) that will receive this event?
```
node.addEventListener('contextmenu', (e) => {
    let fakeEvent = new CustomEvent('mousedown', {
            bubbles: true, 
    	    screenX: e.screenX, 
    	    screenY; e.screenY
            detail: {
                longpressContextMenu: true
            }
        });
     this._handleMouseDown(fakeEvent, index)
}, { passive: true });
```


> If we're going to do it, we should at least construct an actual Event instance (new MouseEvent('mousedown', { ... })) instead of creating an event-like object.

Hm, when I tried it first, `MouseEvent` in `virtualize-table.jsx` was `undefined`... I didn't go into it as I was just trying to get a proof of concept working but yeah, if this is the way we select, it should be a real event.

**Comment by abaevbog** [comment]
Gotcha, apologies, I misunderstood.

So I just tried to just move context-menu logic into a separate listener, and it ... seems to address the issue without breaking anything. 

There are still a few instances when the issue occurs in submodules, specifically the note-editor and the actual reader.

**Comment by abaevbog** [comment]
@mrtcode, #4104 fixed the context menu on long press for the virtualized table and the tab bar. The note editor and the scrollable area  of the reader still have this issue (side bar of the reader opens context menu just fine) - would it make sense for me to try to have a look at those as well?

**Comment by mrtcode** [comment]
@abaevbog Certainly! Just let me know if it will require many changes across the code.

**Comment by abaevbog** [comment]
Cool, I'll give it a shot!

**Comment by abaevbog** [comment]
@mrtcode so these are the tweaks that got the contextmenu to appear on long press for me in reader and note editor.

For note-editor, it was just a matter of swapping events from `mousedown` to `contextmenu`.

For reader, the issue was only present for PDF attachments (snapshots should be ok even now). `contextmenu` opening is handled in `_handlePointerDown`, which gets triggered on long press as if it was just a usual click ( without `event.button === 2`). `_handleContextMenu` listener does fire on the `contextmenu` event, so I just added an extra call to `_handlePointerDown` if that `contextmenu` event comes from a touchscreen, since the `contextmenu` event has the proper `event.button === 2` for the context menu to get opened.

There are aspects of the note-editor and (particularly) the reader's architecture that I am not fully aware of, so correct me if something is missing there or feel free to just use these PRs as a reference point.


https://github.com/zotero/zotero/assets/36271954/c4007fd3-a5d1-436a-a26b-9ad6c11a1085

**Comment by mrtcode** [comment]
Looks good! Merged. Thanks.

**Comment by abaevbog** [comment]
Awesome!

**Comment by dstillman** [comment]
Any more cases of this, or are we good?

**Comment by abaevbog** [comment]
This should be it. The original post was about the context menu not showing in the itemTree, collectionTree (so just virtualized table) and the reader. These were addressed as well as the `note-editor`.
